### PR TITLE
Geolocation association

### PR DESCRIPTION
**Pull request by CoderKevinZhang**
**What type of PR is this?**
> /kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: globalscheduler/controllers/scheduler/scheduler_controller.go [hunk 1]
```diff
@@ -534,3 +548,129 @@ func (sc *SchedulerController) balance() error {
...
+
+func (sc *SchedulerController) balanceScheduler(namespace string) error {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	if len(sc.geoLocationDistribute.CountryMap) == 0 || len(sc.geoLocationDistribute.SchedulersName) == 0 {
+		return nil
+	}
+	schedulersLength := len(sc.geoLocationDistribute.SchedulersName)
+
+	if len(sc.geoLocationDistribute.CountryMap) >= len(sc.geoLocationDistribute.SchedulersName) {
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/scheduler/scheduler_controller.go:573]: you can replace  len(sc.geoLocationDistribute.SchedulersName) with schedulersLength

File: globalscheduler/controllers/util/union/union.go [hunk 2]
```diff
@@ -188,7 +189,8 @@ func unionStorage(unionStorage []*clustercrdv1.StorageSpec, storage []clustercrd
 func unionFlavors(unionFlavors []*clustercrdv1.FlavorInfo, flavors []clustercrdv1.FlavorInfo) []*clustercrdv1.FlavorInfo {
 	if len(unionFlavors) == 0 {
 		for _, y := range flavors {
-			unionFlavors = append(unionFlavors, &y)
+			res := y
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/util/union/union.go:192]: this local variable "res" doesn’t exist after the function’s execution is over.

File: globalscheduler/controllers/scheduler/scheduler_controller.go (hunk 1 above)
Comment (CoderKevinZhang) [reply globalscheduler/controllers/scheduler/scheduler_controller.go]: Done

**Review by jshaofuturewei — APPROVED** [review]

File: globalscheduler/controllers/scheduler/scheduler_controller.go [hunk 3]
```diff
@@ -295,48 +297,57 @@ func (sc *SchedulerController) syncHandler(key *KeyWithEventType) error {
...
 		if err != nil {
-			return fmt.Errorf("cluster object get failed")
+			return err
 		}
-		err = sc.balance()
+
+		sc.geoLocationDistribute.ClustersName = append(sc.geoLocationDistribute.ClustersName, clusterCopy.Name)
+
+		// Store the data according to Cluster geoLocation different fields
+		sc.dataStoreClassified(clusterCopy)
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/scheduler/scheduler_controller.go:325]: could you change the function name to addClusterToGeoMap and change the removeClusterName to removeClusterFromGeoMap

File: globalscheduler/controllers/scheduler/scheduler_controller.go [hunk 4]
```diff
@@ -76,7 +74,8 @@ type SchedulerController struct {
 	clusterInformer   clusterlisters.ClusterLister
 	schedulerSynced   cache.InformerSynced
 
-	consistentHash *consistenthashing.ConsistentHash
+	consistentHash        *consistenthashing.ConsistentHash
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/scheduler/scheduler_controller.go:77]: You can remove the consistent hash field

File: globalscheduler/controllers/scheduler/scheduler_controller.go (hunk 3 above)
Comment (CoderKevinZhang) [reply globalscheduler/controllers/scheduler/scheduler_controller.go]: Done

File: globalscheduler/controllers/scheduler/scheduler_controller.go (hunk 4 above)
Comment (CoderKevinZhang) [reply globalscheduler/controllers/scheduler/scheduler_controller.go]: Done

**Review by cathyhongzhang — APPROVED** [review]
LGTM

**Pull request merged**